internal/clients/team4/speaker.go

```diff
@@ -11,21 +11,67 @@ type speaker struct {
 	parent *client
 }
 
-//s.parent.ServerReadHandle.GetGameConfig()
+//SpeakerActionPriorities indicate speaker actions in order of priority
+var SpeakerActionPriorities = []string{
+	"SetVotingResult",
+	"SetRuleToVote",
+	"AnnounceVotingResult",
+	"UpdateRules",
+	"AppointNextJudge",
+}
+
+func (s *speaker) GetSpeakerBudget() shared.Resources {
+	return s.parent.ServerReadHandle.GetGameState().IIGORolesBudget[shared.Speaker]
+}
+
+func (s *speaker) GetActionsCost(actions []string) shared.Resources {
+	gameconfig := s.parent.ServerReadHandle.GetGameConfig().IIGOClientConfig
+	costs := map[string]shared.Resources{
+		"SetVotingResult":      gameconfig.SetVotingResultActionCost,
+		"SetRuleToVote":        gameconfig.SetRuleToVoteActionCost,
+		"AnnounceVotingResult": gameconfig.AnnounceVotingResultActionCost,
+		"UpdateRules":          gameconfig.UpdateRulesActionCost,
+		"AppointNextJudge":     gameconfig.AppointNextJudgeActionCost,
+	}
+	var SumOfCosts shared.Resources = 0
+	for _, action := range actions {
+		SumOfCosts += costs[action]
+	}
+	return SumOfCosts
+}
+
+func (s *speaker) GetHigherPriorityActionsCost(baseaction string) shared.Resources {
+	actionindex := len(SpeakerActionPriorities)
+	for i, action := range SpeakerActionPriorities {
+		if action == baseaction {
+			actionindex = i
+		}
+	}
+	return s.GetActionsCost(SpeakerActionPriorities[:actionindex])
+}
 
 // PayJudge is used for paying judge for his service
-/*func (s *speaker) PayJudge() shared.SpeakerReturnContent {
+func (s *speaker) PayJudge() shared.SpeakerReturnContent {
 	JudgeSalaryRule, ok := rules.RulesInPlay["salary_cycle_judge"]
 	var JudgeSalary shared.Resources = 0
 	if ok {
 		JudgeSalary = shared.Resources(JudgeSalaryRule.ApplicableMatrix.At(0, 1))
 	}
+	HigherPriorityActionsCost := s.GetHigherPriorityActionsCost("")
+	if s.GetSpeakerBudget()-HigherPriorityActionsCost-JudgeSalary < 0 {
+		//insufficient budget: JudgeSalary + action costs uses up budget
+		JudgeSalary = s.GetSpeakerBudget() - HigherPriorityActionsCost
+		if JudgeSalary < 0 {
+			JudgeSalary = 0
+		}
```

Comment:
JudgeSalary is not deducted from our budget. 
Also, PayJudge is an action that does not deduct from the budget, so there is no need for any budget checking here.

Reply:
Then why does sendJudgeSalary check how much to send from PayJudge? What reason would PayJudge have to alter the amount of the JudgeSalary?

Reply:
Several reasons, but non that I think we need to worry about atm. We could choose to send more or less to the judge for example if the judge is sanctioned, or if we want to be friendly. The flexibility is there for the agents to use if they want to.

Reply:
I think paying more than the salary for the sake of building friendship would be considered nepotism, but ok. For now I'll just reset PayJudge back to how I found it and move on to the other functions.